Building on Arch Linux with QT 6.10.0 fails

Disclaimer: My C++ knowledge is minimal...

While trying to build Z80Explorer on Arch Linux with gcc (version 15.2.1 20251112) and QT 6.10.0 I encountered 2 build errors  in `moc_DialogSchematic.cpp` and `moc_WidgetGraphicsView.cpp` complaining about an incomplete type `struct Trans`. Abridged error from `moc_DialogSchematic.cpp` (`moc_WidgetGraphicsView.cpp` is similiar):
```
In file included from /usr/include/qt6/QtCore/qglobal.h:13,
                 from /usr/include/qt6/QtCore/qnamespace.h:12,
                 from /usr/include/qt6/QtCore/qobjectdefs.h:12,
                 from /usr/include/qt6/QtCore/qobject.h:10,
                 from /usr/include/qt6/QtCore/QObject:1,
                 from ../src/ClassLogic.h:5,
                 from ../src/DialogSchematic.h:5,
                 from moc_DialogSchematic.cpp:9:
/usr/include/c++/15.2.1/type_traits: In instantiation of ‘constexpr const bool std::is_trivially_default_constructible_v<Trans>’:
/usr/include/qt6/QtCore/qtypeinfo.h:33:15:   required from ‘constexpr const bool QtPrivate::qIsComplex<Trans>’
   33 |         !std::is_trivially_default_constructible_v<T> || !std::is_trivially_copyable_v<T>;
      |          ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/qt6/QtCore/qtypeinfo.h:68:32:   required from ‘class QTypeInfo<Trans>’
   68 |         isComplex = QtPrivate::qIsComplex<T>,
      |                     ~~~~~~~~~~~^~~~~~~~~~~~~
/usr/include/qt6/QtCore/qarraydataops.h:827:23:   required by substitution of ‘template<class T> struct QtPrivate::QArrayOpsSelector<T, typename std::enable_if<(QTypeInfo<T>::isComplex && QTypeInfo<T>::isRelocatable)>::type> [with T = Trans]’
  827 |         QTypeInfo<T>::isComplex && QTypeInfo<T>::isRelocatable
      |                       ^~~~~~~~~
/usr/include/qt6/QtCore/qarraydataops.h:834:8:   required from ‘struct QtPrivate::QCommonArrayOps<Trans>’
  834 | struct QCommonArrayOps : QArrayOpsSelector<T>::Type
      |        ^~~~~~~~~~~~~~~
/usr/include/qt6/QtCore/qarraydataops.h:918:8:   required from ‘struct QArrayDataOps<Trans>’
  918 | struct QArrayDataOps
      |        ^~~~~~~~~~~~~
/usr/include/qt6/QtCore/qarraydatapointer.h:109:22:   required from ‘QArrayDataPointer<T>::~QArrayDataPointer() [with T = Trans]’
  109 |             (*this)->destroyAll();
      |             ~~~~~~~~~^~~~~~~~~~
/usr/include/qt6/QtCore/qlist.h:300:15:   required from here
  300 |     constexpr QList() noexcept = default;
      |               ^~~~~
/usr/include/c++/15.2.1/type_traits:3584:7: error: invalid use of incomplete type ‘struct Trans’
 3584 |     = __is_trivially_constructible(_Tp);
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/ClassLogic.h:7:8: note: forward declaration of ‘struct Trans’
    7 | struct Trans;
      |        ^~~~~
...
```

I added `#include "ClassNetlist.h"` to `DialogSchematic.h` and `WidgetGraphicsView.h` to get the `struct Trans` definition:
```
diff --git a/src/DialogSchematic.h b/src/DialogSchematic.h
index 1f85798..39e043b 100644
--- a/src/DialogSchematic.h
+++ b/src/DialogSchematic.h
@@ -3,6 +3,7 @@
 
 #include "AppTypes.h"
 #include "ClassLogic.h"
+#include "ClassNetlist.h"
 #include <QDialog>
 #include <QGraphicsScene>
 #include <QMenu>
diff --git a/src/WidgetGraphicsView.h b/src/WidgetGraphicsView.h
index cc79229..d63d1f0 100644
--- a/src/WidgetGraphicsView.h
+++ b/src/WidgetGraphicsView.h
@@ -2,6 +2,7 @@
 #define WIDGETGRAPHICSVIEW_H
 
 #include "ClassLogic.h"
+#include "ClassNetlist.h"
 #include <QGraphicsPolygonItem>
 #include <QGraphicsView>
```

I also got 2 warnings from gcc:

1.  In `ClassLogic.cpp` gcc complains about initializer list ordering:

```
In file included from ../src/ClassNetlist.h:5,
                 from ../src/ClassSimZ80.h:4,
                 from ../src/ClassController.h:9,
                 from ../src/ClassLogic.cpp:1:
../src/ClassLogic.h: In constructor ‘Logic::Logic(net_t, LogicOp, bool, bool)’:
../src/ClassLogic.h:26:11: warning: ‘Logic::outnet’ will be initialized after [-Wreorder]
   26 |     net_t outnet;                       // The output net number, or the net itself (for LogicOp::Net)
      |           ^~~~~~
../src/ClassLogic.h:20:13: warning:   ‘LogicOp Logic::op’ [-Wreorder]
   20 |     LogicOp op;                         // Specifies the logic operation on its input nodes
      |             ^~
../src/ClassLogic.cpp:18:1: warning:   when initialized here [-Wreorder]
   18 | Logic::Logic(net_t n, LogicOp op, bool checkVisitedNets, bool first) : outnet(n), op(op)
      | ^~~~~
```

I changed the order:
```
diff --git a/src/ClassLogic.cpp b/src/ClassLogic.cpp
index 5733b9b..1a27d8f 100644
--- a/src/ClassLogic.cpp
+++ b/src/ClassLogic.cpp
@@ -15,7 +15,7 @@
 static QVector<net_t> visitedNets; // Avoid loops by keeping nets that are already visited
 static QVector<tran_t> visitedTrans; // Avoid path duplication by keeping transistors that are already visited
 
-Logic::Logic(net_t n, LogicOp op, bool checkVisitedNets, bool first) : outnet(n), op(op)
+Logic::Logic(net_t n, LogicOp op, bool checkVisitedNets, bool first) : op(op), outnet(n)
 {
     QSettings settings;
     QString termNodes = settings.value("schematicTermNodes").toString();
```

2. In `ClassVisual.cpp` gcc complains about an uninitialized variable:
```
../src/ClassVisual.cpp: In member function ‘void ClassVisual::drawNets(QPainter&, const QRect&, bool, uint)’:
../src/ClassVisual.cpp:822:13: warning: ‘active’ may be used uninitialized [-Wmaybe-uninitialized]
  822 |             if (active)
      |             ^~
../src/ClassVisual.cpp:805:18: note: ‘active’ was declared here
  805 |             bool active;
      |                  ^~~~~~
```

I initialized it:
```
diff --git a/src/ClassVisual.cpp b/src/ClassVisual.cpp
index 7762a3f..ccc373b 100644
--- a/src/ClassVisual.cpp
+++ b/src/ClassVisual.cpp
@@ -802,7 +802,7 @@ void ClassVisual::drawNets(QPainter &painter, const QRect &viewport, bool order,
     {
         if (!::controller.getSimZ80().isNetOrphan(i))
         {
-            bool active;
+            bool active = false;
             switch (mode)
             {
                 case 0: // 0:Active
```

These changes gave me a clean compile.

--- a/src/ClassLogic.cpp
+++ b/src/ClassLogic.cpp
@@ -15,7 +15,7 @@
 static QVector<net_t> visitedNets; // Avoid loops by keeping nets that are already visited
 static QVector<tran_t> visitedTrans; // Avoid path duplication by keeping transistors that are already visited
 
-Logic::Logic(net_t n, LogicOp op, bool checkVisitedNets, bool first) : outnet(n), op(op)
+Logic::Logic(net_t n, LogicOp op, bool checkVisitedNets, bool first) : op(op), outnet(n)
 {
     QSettings settings;
     QString termNodes = settings.value("schematicTermNodes").toString();

--- a/src/ClassVisual.cpp
+++ b/src/ClassVisual.cpp
@@ -802,7 +802,7 @@ void ClassVisual::drawNets(QPainter &painter, const QRect &viewport, bool order,
     {
         if (!::controller.getSimZ80().isNetOrphan(i))
         {
-            bool active;
+            bool active = false;
             switch (mode)
             {
                 case 0: // 0:Active

--- a/src/DialogSchematic.h
+++ b/src/DialogSchematic.h
@@ -3,6 +3,7 @@
 
 #include "AppTypes.h"
 #include "ClassLogic.h"
+#include "ClassNetlist.h"
 #include <QDialog>
 #include <QGraphicsScene>
 #include <QMenu>

--- a/src/WidgetGraphicsView.h
+++ b/src/WidgetGraphicsView.h
@@ -2,6 +2,7 @@
 #define WIDGETGRAPHICSVIEW_H
 
 #include "ClassLogic.h"
+#include "ClassNetlist.h"
 #include <QGraphicsPolygonItem>
 #include <QGraphicsView>
 

